docs/adr/0008-allowing-more-types-as-errors.md

```diff
@@ -0,0 +1,139 @@
+impl ErrorDetails {
+    fn message(&self) -> String { todo!() }
+}
+fn do_it() -> Result<(), ErrorDetails> { Err(ErrorDetails {}) }
+```
+
+How should this be captured on the foreign side?
+
+### A) Create an "error only" type
+
+This is like the status quo re enums. It means:
+* There's no "non error" version of the interface.
+* A struct simply appearing as an error changes how the foreign side sees that struct significantly.
+* The type can't be used exactly as it could in non-error contexts.
+
+ie, it's a bit of "either or, but not both". But it makes for more natural foreign code:
+
+```
+try:
+    do_it()
+except ErrorDetailsException as e:
+    # There's no `ErrorDetails` type in this world.
+    # ie, `e' is an  `ErrorDetailsException`
+    # `ErrorDetailsException` derives from `Exception`
+    e.message()
+```
+
+### B) Create an original type and an error type
+
+We create 2 types - the `interface` itself unchanged from how the struct would be generated now,
+and another exception "wrapper" for the type.
+It means you *catch* a slightly different name from the interface, but the non-error version of the interface is what you deal with.
+
+```
+try:
+    do_it()
+except ErrorDetailsException as e:
+    # There are 2 types here:
+    # `e` is an `ErrorDetailsException`
+    # `e.inner` (or any other name) is an `ErrorDetails` - does not derive from `Exception`
+    # We can make it look natural via delegation.
+    e.message()
+    # But also use the original interface
+    HandleErrorReportingProblem(e.inner)
+```
+
+To clarify, you could imagine generated Python code looking something like:
+```
+class ErrorDetails:
+    ...
+class ErrorDetailsException(Exception):
+    ...
+```
+### C) Create and original type with an error type nested within the original
+
+Very similar to the above - there are still 2 unique types created, but only one type is a top-level
+type - the error type is created inside the exception.
+
+The code catching exceptions might look something like:
+```
+try:
+    do_it()
+except ErrorDetails.Exception as e:
+    # same block above as for B
+```
+
+ie, there's no top-level `ErrorDetailsException` type, it's accessed via `ErrorDetails.Exception`
+
+To clarify, you could imagine generated Python code looking something like:
+```
+class ErrorDetails:
+    # snip regular interface.
+    class Exception(Exception):
+        # snip error specific code
+```
+
+## Decision Outcome
+
+Chosen option: "[option X]", because [justification. e.g., only option, which meets k.o. criterion decision driver | which resolves force force | … | comes out best (see below)].
+
+## Pros and Cons of the Options
+
+### 1: Do nothing - don't support interfaces as errors.
+
+Con: Seems a bit negative :)
+
+### 2: Implement Option A above
+
+- Pro: no magic, there's exactly 1 type for the type in the UDL/macro
+- Con: That type gets "tainted" by being used as an error. Using the interface in other contexts is less natural.
```

Comment:
I think this is both pro and con.  It's unnatural from the Rust POV, but natural from the foreign binding POV.

Reply:
I think it depends on the context: today, if I see a `struct Foo` in Rust, I know how that's going to be represented by UniFFI in bindings. With option A, that's no longer true - I now need to search for every single place `Foo` is used, in all crates, and if I can identify even a single one as being an `E` in a `Result<>`, then that representation changes. I think that's really my biggest objection to A - that how `Foo` is used in a place which is far far away from where `Foo` is defined changes what you understand about `Foo` in a significant and global way.

Reply:
If we still require an `[Error]` annotation on that type it's clear, isn't it?

Reply:
We don't need an `[Error]` annotation. I guess we could insist on one anyway, but that is a bit of a smell for me - ie, with (a) "we would insist on an error attribute so that other side-effects (ie, the fact this is used somewhere as an error causes the object to change significantly) become less surprising" vs (b) and (c) "no side effects so no need for additional annotations"